### ISSUE TITLE
sqlfluff cannot parse more complicated array slicing for PostgreSQL

### ISSUE THREAD

**Issue by jonasbb** [sqlfluff/sqlfluff#4336]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

```sql
SELECT total[times.time_start :times.time_start + times.train_window - 1] total
FROM
    my_table,
    times
```

Linting the  above SQL fails with the following error:
```
L:   1 | P:  13 |  PRS | Line 1, Position 13: Found unparsable section:
                       | '[times.time_start :times.time_start + ti...'
WARNING: Parsing errors found and dialect is set to 'postgres'. Have you configured your dialect correctly?
```

I actually found two related issues #3842 and #3903, the second being the implementation closing the first one. Reading #3903 the problem might be, that sqlfluff only allows for numeric literals, but not column names and expressions in the array slicing.

### Expected Behaviour

No error should occur, and running `sqlfluff fix` should fix the file.
The SQL works fine when used with PostgreSQL.

### Observed Behaviour

L:   1 | P:  13 |  PRS | Line 1, Position 13: Found unparsable section:
                       | '[times.time_start :times.time_start + ti...'
WARNING: Parsing errors found and dialect is set to 'postgres'. Have you configured your dialect correctly?

### How to reproduce

`sqlfluff lint --dialect postgres ./sqlfluff-bug.sql`

`./sqlfluff-bug.sql`:
```sql
SELECT total[times.time_start :times.time_start + times.train_window - 1] total
FROM
    my_table,
    times
```

### Dialect

postgres

### Version

sqlfluff, version 1.4.5
Python 3.11.1

### Configuration

# No configuration file

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by james-johnston-thumbtack** [comment]
Here's another example which demonstrates two different failures to parse:

```sql
-- Based on https://wiki.postgresql.org/wiki/Unindexed_foreign_keys
SELECT
    quote_ident(pg_namespace.nspname) || '.' || quote_ident(pg_class.relname) AS referencing_tbl,
    quote_ident(pg_attribute.attname) AS referencing_column
FROM pg_constraint
INNER JOIN
    pg_attribute
    ON
        pg_attribute.attrelid = pg_constraint.conrelid
        AND pg_attribute.attnum = pg_constraint.conkey[1]
INNER JOIN pg_class ON pg_class.oid = pg_constraint.conrelid
INNER JOIN pg_namespace ON pg_namespace.oid = pg_class.relnamespace
WHERE
    pg_constraint.contype = 'f'
    AND NOT EXISTS (
        SELECT 1
        FROM pg_index AS i
        WHERE
            i.indrelid = pg_constraint.conrelid
            AND ((i.indkey::int2[])[:])[1:cardinality(pg_constraint.conkey)] @> pg_constraint.conkey
    )
ORDER BY referencing_tbl, referencing_column;
```

Results:

```
% sqlfluff lint .
== [unindexed_foreign_keys.sql] FAIL
L:  20 | P:  36 |  PRS | Line 20, Position 36: Found unparsable section: '[:]'
L:  20 | P:  40 |  PRS | Line 20, Position 40: Found unparsable section:
                       | '[1:cardinality(pg_constraint.conkey)] @>...'
WARNING: Parsing errors found and dialect is set to 'postgres'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

Note that this SQL works fine.

So, there are two problems:

- https://github.com/sqlfluff/sqlfluff/pull/3903 apparently handled the case when either the lower bound or the upper bound was missing.  However, it does not correctly handle the case when BOTH are missing.  This could be used to reset the array bounds as described at https://stackoverflow.com/a/44448230
- As noted earlier, it can't seem to parse more complex expressions than a simple numeric literal.
